### PR TITLE
[Woo Pos] Allow adding products to Cart

### LINKED ISSUE #11651: [Woo POS] Sync product selection state between Product Selector and Cart

### PR DESCRIPTION
**Pull request by samiuelson**
Closes: #11651 

### Description

This PR introduces the following functionalities:
* Adding items to the cart by clicking products in product selector
* Removing items from the cart by clicking "X"

### Steps to reproduce

* Add and remove items to cart
* Verify items are visible in the checkout as well
* Test that the state is correctly preserved during configuration changes

### Images/gif

https://github.com/woocommerce/woocommerce-android/assets/4527432/ca13b101-8302-4951-9b6e-ce6bc9b2e9ac

- [x] I have considered if this change warrants release notes and have added them to `RELEASE-NOTES.txt` if necessary. Use the "[Internal]" label for non-user-facing changes.

### CODE REVIEW

**Comment by dangermattic** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/products/WooPosProductsViewModel.kt [hunk 1]
```diff
@@ -35,10 +41,42 @@ class WooPosProductsViewModel @Inject constructor(
...
+    )
+
     fun onEndOfProductsGridReached() {
         loadMoreProductsJob?.cancel()
         loadMoreProductsJob = launch {
             productsDataSource.loadMore()
         }
     }
+
+    fun onItemClicked(itemClicked: WooPosProductsListItem) {
```
Comment (kidinov) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/products/WooPosProductsViewModel.kt]: My understanding of the flow was that on the products side you can only add products to cart. Removing of products possible only from the cart

With the current implementation how can you add 2 or more products?

If my understanding is correct that we won't keep selected products in the products VM, only in the cart. And the only event that has to be sent from Product to Cart is ID of a selected product

Comment (samiuelson) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/products/WooPosProductsViewModel.kt]: Good point. I simplified the implementation and removed the "selected" state from the products pane. This allowed adding multiple same products – [e2ae955](https://github.com/woocommerce/woocommerce-android/pull/11699/commits/e2ae95546d5166cc877e385241fb6804f09f0a62)
Also, added possibility to remove items from cart – [1e622a8](https://github.com/woocommerce/woocommerce-android/pull/11699/commits/1e622a8821cf91885f64244a481ea170adcb1c20)

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/WooPosHomeChildToParentCommunication.kt [hunk 2]
```diff
@@ -20,10 +21,13 @@ class WooPosChildrenToParentCommunication @Inject constructor() :
 sealed class ChildToParentEvent {
     data object CheckoutClicked : ChildToParentEvent()
     data object BackFromCheckoutToCartClicked : ChildToParentEvent()
+    data class ItemClickedInProductSelector(
+        val selectedItem: WooPosProductsListItem
```
Comment (kidinov) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/WooPosHomeChildToParentCommunication.kt]: I think we should pass only the product ID here as:
* We are more likely to not hit really hard to fix crashes with [TransactionTooLargeException](https://developer.android.com/reference/android/os/TransactionTooLargeException) - you store objects list in the bundle
* We kinda attach Products and a Cart together. Now products kinda dictate what UI VM you use in the cart

And overall I think reusing the same UI VM (`WooPosProductsListItem`) for 2 screens is not really a good idea, especially when don't know what data we are going to show on each of those

There is `getProductByRemoteId` in `ProductStore` to get a product that you can already map to the `WooPosCartListItem`

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/products/WooPosProductsScreen.kt [hunk 3]
```diff
@@ -41,13 +42,15 @@ fun WooPosProductsScreen() {
     val productsViewModel: WooPosProductsViewModel = hiltViewModel()
     WooPosProductsScreen(
         productsState = productsViewModel.viewState,
+        onItemClicked = productsViewModel::onItemClicked,
```
Comment (kidinov) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/products/WooPosProductsScreen.kt]: Shell we use "events", e.g. as `fun onUIEvent(event: WooPosCartUIEvent) {` in the cart but  but for products?
* For consistency
* No need to pass many lyambdas and add them for every new event to all the layers 
* We have compile time defined set of UI events, so when a new one is added - compile will tell as where we must handle it

Here we already have `OnProductsListEndReached` and `OnProductClicked`

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/cart/WooPosCartViewModel.kt [hunk 4]
```diff
@@ -28,12 +35,26 @@ class WooPosCartViewModel @Inject constructor(
...
-                _state.value = WooPosCartState.Checkout
+                _state.update { state ->
+                    WooPosCartState.Checkout(state.itemsInCart)
+                }
             }
 
             is WooPosCartUIEvent.BackFromCheckoutToCartClicked -> {
                 sendEventToParent(ChildToParentEvent.BackFromCheckoutToCartClicked)
-                _state.value = WooPosCartState.Cart
+                _state.update { state ->
```
Comment (kidinov) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/cart/WooPosCartViewModel.kt]: I look into this update function:

```
public inline fun <T> MutableStateFlow<T>.update(function: (T) -> T) {
    while (true) {
        val prevValue = value
        val nextValue = function(prevValue)
        if (compareAndSet(prevValue, nextValue)) {
            return
        }
    }
}
```

and can not figure out why we need to use it here?

**Review by kidinov — COMMENTED** [review]
LGTM! I left a couple of suggestions, please take a look

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/products/WooPosProductsScreen.kt (hunk 3 above)
Comment (samiuelson) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/products/WooPosProductsScreen.kt]: Done – aff5980441315fc4851c8da27942d5733a86d32d

**Comment by dangermattic** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Class <code>WooPosProductsUIEvent</code> is missing tests, but <code>unit-tests-exemption</code> label was set to ignore this.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/WooPosHomeChildToParentCommunication.kt (hunk 2 above)
Comment (samiuelson) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/WooPosHomeChildToParentCommunication.kt]: Good point. Updated – e7c46de40ba0baa0387c2e7bba23ac65da0cb6a8

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/cart/WooPosCartViewModel.kt (hunk 4 above)
Comment (samiuelson) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/cart/WooPosCartViewModel.kt]: I switched to `flow.value` syntax – bd1ab06465a0ab96569b2d102de318fff127bc07

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/cart/WooPosCartViewModel.kt [hunk 5]
```diff
@@ -43,7 +66,23 @@ class WooPosCartViewModel @Inject constructor(
             parentToChildrenEventReceiver.events.collect { event ->
                 when (event) {
                     is ParentToChildrenEvent.BackFromCheckoutToCartClicked -> {
-                        _state.value = WooPosCartState.Cart
+                        _state.value = WooPosCartState.Cart(state.value.itemsInCart)
+                    }
+
+                    is ParentToChildrenEvent.ItemClickedInProductSelector -> {
+                        _state.update { state ->
+                            val site = site.getOrNull() ?: return@collect
```
Comment (kidinov) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/cart/WooPosCartViewModel.kt]: Wdyt about crashing in both cases here, as both cases mean that we are in unrecoverable state anyway already?


```
viewModelScope.async {
                                val itemClicked = 
                                    productStore.getProductByRemoteId(site.get(), event.productId)!!.toCartListItem()
                                val itemsInCart = state.itemsInCart + itemClicked

                                when (state) {
                                    is WooPosCartState.Cart -> state.copy(itemsInCart = itemsInCart)
                                    is WooPosCartState.Checkout -> state.copy(itemsInCart = itemsInCart)
                                }
                            }.await()
```

**Review by kidinov — APPROVED** [review]
LGTM! I left probably non-blocking, but in my opinion important comment. Please take a look

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/cart/WooPosCartViewModel.kt (hunk 5 above)
Comment (samiuelson) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/cart/WooPosCartViewModel.kt]: Makes sense – [b4f7c92](https://github.com/woocommerce/woocommerce-android/pull/11699/commits/b4f7c929eed835edf7e83bdd3e36dcb3af693a0e)

**Pull request merged**